### PR TITLE
Fix doc equality

### PR DESCRIPTION
**Pull request by davidbp**
Current equality of documents provides unnexpected behaviours, such as two documents one with embedding and the other without embedding beeing equal. Same happens with tensor.

Example

```python
print('same document')
d1 = Document(id=23, embedding=np.array([1,2,3]))
d2 = Document(id=23, embedding=np.array([1,2,3]))
print('d1 == d2             --->',d1 == d2)
print('d1._data == d2._data --->',d1._data == d2._data)
print('d1.non_empty_fields  --->',d1.non_empty_fields)
print('d2.non_empty_fields  --->',d2.non_empty_fields)

print('\nDocument with and without embedding')
d1 = Document(id=23, embedding=np.array([1,2,3]))
d2 = Document(id=23)
print('d1 == d2             --->',d1 == d2)
print('d1._data == d2._data --->',d1._data == d2._data)
print('d1.non_empty_fields  --->',d1.non_empty_fields)
print('d2.non_empty_fields  --->',d2.non_empty_fields)

print('\nDocument with different  embeddings')
d1 = Document(id=23, embedding=np.array([1,2,3]))
d2 = Document(id=23, embedding=np.array([1,2,4]))
print('d1 == d2             --->',d1 == d2)
print('d1._data == d2._data --->',d1._data == d2._data)
print('d1.non_empty_fields  --->',d1.non_empty_fields)
print('d2.non_empty_fields  --->',d2.non_empty_fields)
```

In main this produces 

```
same document
d1 == d2             ---> True
d1._data == d2._data ---> True
d1.non_empty_fields  ---> ('id', 'embedding')
d2.non_empty_fields  ---> ('id', 'embedding')

Document with and without embedding
d1 == d2             ---> True
d1._data == d2._data ---> True
d1.non_empty_fields  ---> ('id', 'embedding')
d2.non_empty_fields  ---> ('id',)

Document with different  embeddings
d1 == d2             ---> True
d1._data == d2._data ---> True
d1.non_empty_fields  ---> ('id', 'embedding')
d2.non_empty_fields  ---> ('id', 'embedding')
```

In this PR

```
same document
d1 == d2             ---> True
d1._data == d2._data ---> True
d1.non_empty_fields  ---> ('id', 'embedding')
d2.non_empty_fields  ---> ('id', 'embedding')

Document with and without embedding
d1 == d2             ---> False
d1._data == d2._data ---> False
d1.non_empty_fields  ---> ('id', 'embedding')
d2.non_empty_fields  ---> ('id',)

Document with different  embeddings
d1 == d2             ---> False
d1._data == d2._data ---> False
d1.non_empty_fields  ---> ('id', 'embedding')
d2.non_empty_fields  ---> ('id', 'embedding')
```

### CODE REVIEW

File: docarray/document/data.py [hunk 1]
```diff
@@ -111,3 +113,25 @@ def _set_default_value_if_none(self, key):
                     setattr(self, key, defaultdict(NamedScore))
                 else:
                     setattr(self, key, v() if callable(v) else v)
+
+    @staticmethod
+    def _embedding_eq(array1: 'ArrayType', array2: 'ArrayType'):
```
Comment (alaeddine-13) [inline docarray/document/data.py:118]: I guess a similar method needs to be implemented for tensor

Comment (davidbp) [reply docarray/document/data.py:119]: you are right, added

**Comment by hanxiao** [comment]
btw the behavior is due to that i intentionally added `hash=False` to some fields in `class DocumentData:`,

File: docarray/math/ndarray.py [hunk 2]
```diff
@@ -192,3 +192,57 @@ def get_array_rows(array: 'ArrayType') -> Tuple[int, int]:
...
+                same_array = (x - y).sum() == 0
+            elif x_type == 'tensorflow':
+                if x_is_sparse:
+                    same_array = x == y
+                else:
+                    same_array = (x == y).numpy().all()
+            elif x_type == 'paddle':
+                # Paddle does not support sparse tensor on 11/8/2021
+                # https://github.com/PaddlePaddle/Paddle/issues/36697
+                same_array = (x == y).numpy().all()
```
Comment (hanxiao) [inline docarray/math/ndarray.py:244]: im not so sure about this logic here, shouldn't we use package-specific assert equal there?

e.g. 

- for torch: `torch.eq(a, b)`
- for numpy: `numpy.array_equal`
- for scipy: google "test two scipy matrix are equal"
- for tf
- ...

**Review by hanxiao — CHANGES_REQUESTED** [review]
i believe a better way is to use package specific API instead of writing it by ourselves

File: docarray/document/data.py [hunk 3]
```diff
@@ -111,3 +113,37 @@ def _set_default_value_if_none(self, key):
...
+            return check_arraylike_equality(array1, array2)
+        else:
+            return False
+
+    @staticmethod
+    def _tensor_eq(array1: 'ArrayType', array2: 'ArrayType'):
+        DocumentData._embedding_eq(array1, array2)
+
+    def __eq__(self, other):
+        for key in self.__dataclass_fields__.keys():
```
Comment (hanxiao) [inline docarray/document/data.py:133]: do not use this, use `non_empty_fields` or `_non_empty_fields`

**Pull request merged**